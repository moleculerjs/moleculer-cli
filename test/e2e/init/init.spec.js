import { describe, it, expect, beforeEach } from "vitest";
import { execa } from "execa";

const binPath = "./bin/moleculer.js";

const rm = require("rimraf");
const fs = require("fs");

const TEMPLATE_ROOT = "./test/e2e/init/templates/";
const TMP_PATH = "./tmp";

describe("E2E: init command", () => {
	describe("Init 'no-meta' template", () => {
		const templateName = "no-meta";
		const templatePath = TEMPLATE_ROOT + "/" + templateName;

		beforeEach(() => rm.sync(TMP_PATH));

		it("should generate template", async () => {
			const { stdout, stderr, exitCode } =
				await execa`node ${binPath} init ${templatePath} e2e-demo ./tmp --no-install`;

			console.log(stdout);
			expect(stdout).toContain("Done!");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/e2e-demo.project", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/index.js", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
		}, 10_000);
	});

	describe("Init 'handlebars' template", () => {
		const templateName = "handlebars";
		const templatePath = TEMPLATE_ROOT + "/" + templateName;

		beforeEach(() => rm.sync(TMP_PATH));

		it("should generate template with lint file", async () => {
			const { stdout, stderr, exitCode } = await execa("node", [
				binPath,
				"init",
				templatePath,
				"e2e-demo",
				"./tmp",
				"--no-install",
				"--@fullName=John Doe",
				"--@projectDescription='Test project'",
				"--@lint=true"
			]);

			console.log(stdout);
			expect(stdout).toContain("Before hook");
			expect(stdout).toContain("After hook");
			expect(stdout).toContain("Complete hook");
			expect(stdout).toContain("Created project 'e2e-demo'!");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/e2e-demo.project", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/index.js", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/SKIP.md", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/.eslintrc.js", "utf8")).toMatchSnapshot();
		}, 10_000);

		it("should generate template without lint file", async () => {
			const { stdout, stderr, exitCode } = await execa("node", [
				binPath,
				"init",
				templatePath,
				"e2e-demo",
				"./tmp",
				"--no-install",
				"--@fullName=John Doe",
				"--@projectDescription='Test project'",
				"--@lint=false"
			]);

			console.log(stdout);
			expect(stdout).toContain("Before hook");
			expect(stdout).toContain("After hook");
			expect(stdout).toContain("Complete hook");
			expect(stdout).toContain("Created project 'e2e-demo'!");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/e2e-demo.project", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/index.js", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/SKIP.md", "utf8")).toMatchSnapshot();
			expect(fs.existsSync(TMP_PATH + "/.eslintrc.js")).toBe(false);
		}, 10_000);
	});

	describe("Init 'nunjucks' template", () => {
		const templateName = "nunjucks";
		const templatePath = TEMPLATE_ROOT + "/" + templateName;

		beforeEach(() => rm.sync(TMP_PATH));

		it("should generate template with lint file", async () => {
			const { stdout, stderr, exitCode } = await execa("node", [
				binPath,
				"init",
				templatePath,
				"e2e-demo",
				"./tmp",
				"--no-install",
				"--@fullName=John Doe",
				"--@projectDescription='Test project'",
				"--@lint=true"
			]);

			console.log(stdout);
			expect(stdout).toContain("Done!");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/e2e-demo.project", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/package.json", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
		}, 10_000);

		it("should generate template without lint file", async () => {
			const { stdout, stderr, exitCode } = await execa("node", [
				binPath,
				"init",
				templatePath,
				"e2e-demo",
				"./tmp",
				"--no-install",
				"--@fullName=John Doe",
				"--@projectDescription='Test project'",
				"--@lint=false"
			]);

			console.log(stdout);
			expect(stdout).toContain("Done!");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/e2e-demo.project", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/package.json", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
		}, 10_000);
	});

	describe("Init 'custom-renderer' template", () => {
		const templateName = "custom-renderer";
		const templatePath = TEMPLATE_ROOT + "/" + templateName;

		beforeEach(() => rm.sync(TMP_PATH));

		it("should register as alias", async () => {
			const { stdout, stderr, exitCode } =
				await execa`node ${binPath} alias-template --force asd ${templatePath}`;

			console.log(stdout);
			expect(stdout).toContain("Template alias 'asd' with value");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");
		});

		it("should generate template with custom renderer", async () => {
			const { stdout, stderr, exitCode } = await execa("node", [
				binPath,
				"init",
				"asd",
				"e2e-demo",
				"./tmp",
				"--no-install",
				"--@fullName=Jane Doe",
				"--@projectDescription='Test project'",
				"--@lint=true"
			]);

			console.log(stdout);
			expect(stdout).toContain("Done!");
			expect(stdout).toContain("Rendering file with Mustache...");
			expect(stdout).toContain("id: 306521");
			expect(exitCode).toBe(0);
			expect(stderr).toBe("");

			expect(fs.readFileSync(TMP_PATH + "/index.js", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/package.json", "utf8")).toMatchSnapshot();
			expect(fs.readFileSync(TMP_PATH + "/README.md", "utf8")).toMatchSnapshot();
		}, 10_000);
	});
});
