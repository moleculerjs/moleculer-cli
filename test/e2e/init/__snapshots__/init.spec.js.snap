// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`E2E: init command > Init 'custom-renderer' template > should generate template with custom renderer 1`] = `
"/*
 * e2e-demo
 * Copyright (c) 2025 Jane Doe (https://github.com//e2e-demo)
 * MIT Licensed
 */

module.exports = require("./src");
"
`;

exports[`E2E: init command > Init 'custom-renderer' template > should generate template with custom renderer 2`] = `
"{
  "name": "e2e-demo",
  "version": "0.1.0",
  "description": "Test project",
  "main": "index.js",
  "scripts": {
    "demo": "node examples/index.js",
    "lint": "eslint --ext=.js src test",
    "deps": "npm-check -u"
  },
  "author": "Jane Doe",
  "license": "MIT",
  "devDependencies": {
    "eslint": "3.19.0",
    "jest": "20.0.0",
    "jest-cli": "20.0.0"
  }
}
"
`;

exports[`E2E: init command > Init 'custom-renderer' template > should generate template with custom renderer 3`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)

Test project

## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 Jane Doe
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template with lint file 1`] = `
"e2e-demo
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template with lint file 2`] = `
"/*
 * e2e-demo
 * Copyright (c) 2025 John Doe (https://github.com//e2e-demo)
 * MIT Licensed
 */

module.exports = require("./src");
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template with lint file 3`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)

Test project

## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 John Doe
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template with lint file 4`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# {{projectName}} [![NPM version](https://img.shields.io/npm/v/{{projectName}}.svg)](https://www.npmjs.com/package/{{projectName}})

{{projectDescription}}

## Features

## Install
\`\`\`
npm install {{projectName}} --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) {{year}} {{fullName}}
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template with lint file 5`] = `
"module.exports = {
	root: true,

	rules: {
		indent: ["warn", "tab"]
	}
};
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template without lint file 1`] = `
"e2e-demo
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template without lint file 2`] = `
"/*
 * e2e-demo
 * Copyright (c) 2025 John Doe (https://github.com//e2e-demo)
 * MIT Licensed
 */

module.exports = require("./src");
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template without lint file 3`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)

Test project

## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 John Doe
"
`;

exports[`E2E: init command > Init 'handlebars' template > should generate template without lint file 4`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# {{projectName}} [![NPM version](https://img.shields.io/npm/v/{{projectName}}.svg)](https://www.npmjs.com/package/{{projectName}})

{{projectDescription}}

## Features

## Install
\`\`\`
npm install {{projectName}} --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) {{year}} {{fullName}}
"
`;

exports[`E2E: init command > Init 'no-meta' template > should generate template 1`] = `
"e2e-demo
"
`;

exports[`E2E: init command > Init 'no-meta' template > should generate template 2`] = `
"/*
 * e2e-demo
 * Copyright (c) 2025 
 * MIT Licensed
 */

module.exports = require("./src");
"
`;

exports[`E2E: init command > Init 'no-meta' template > should generate template 3`] = `
"# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)



## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template with lint file 1`] = `
"e2e-demo
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template with lint file 2`] = `
"{
  "name": "e2e-demo",
  "version": "0.1.0",
  "description": "Test project",
  "main": "index.js",
  "scripts": {
    "demo": "node examples/index.js","lint": "eslint --ext=.js src test",
    "deps": "npm-check -u"
  },
  "author": "John Doe",
  "license": "MIT",
  "devDependencies": {"eslint": "3.19.0",
    "jest": "20.0.0",
    "jest-cli": "20.0.0"
  }
}
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template with lint file 3`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)

Test project

## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 John Doe
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template without lint file 1`] = `
"e2e-demo
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template without lint file 2`] = `
"{
  "name": "e2e-demo",
  "version": "0.1.0",
  "description": "Test project",
  "main": "index.js",
  "scripts": {
    "demo": "node examples/index.js",
    "deps": "npm-check -u"
  },
  "author": "John Doe",
  "license": "MIT",
  "devDependencies": {
    "jest": "20.0.0",
    "jest-cli": "20.0.0"
  }
}
"
`;

exports[`E2E: init command > Init 'nunjucks' template > should generate template without lint file 3`] = `
"![Moleculer logo](http://moleculer.services/images/banner.png)

# e2e-demo [![NPM version](https://img.shields.io/npm/v/e2e-demo.svg)](https://www.npmjs.com/package/e2e-demo)

Test project

## Features

## Install
\`\`\`
npm install e2e-demo --save
\`\`\`


# License
The project is available under the [MIT license](https://tldrlegal.com/license/mit-license).

# Contact
Copyright (c) 2025 John Doe
"
`;
